New approach

aleixjf opened the pull request:
Hi Marekkon,

After learning the RUST basics, and fighting resolving all the Ownership and Move errors (WOW... What a headache) I've managed to update part of the src code.  Currently, the app builds correctly, but the auto tag feature doesn't seem to work. However, I'm pulling this request because you'll surely identify the reason why it's not working and can take advantage of this "new approach".

The idea is that there is a single "Track" structure, with all its fields as Optionnals, so that by default they are set to None.
When building a new Track object, either from the API or from the local metadata, you provide all the info you have, and once created, you trigger a "fill_tags" function (defined in the Track impl), which goes through the available fields and fills the empty ones, if possible.
Like this, all API tracks are built similarly to the way the local track object is built, and it's easier to compare (you'll find the version in a separate matcher.rs file, inside the tagger folder), and we should be able to get even better successful match results.
If matched, you then get returned the API track object, and you fill the local track metadata with the API track object.

As I said, the app is building correctly, but the Auto Tag process isn't working (it returns a 100% complete, with 0 tracks matched/skipped/with error), and I'm unable to debug it. I've opened the DevTools in order to check in the console what messages were sent, but it said that DevTools couldn't load the origin maps.

I'll be more than happy to help and contribute to this project! :)

Marekkon5 commented:
Hello, I really appreciate the effort you put into this, however I am afraid I won't be merging this, as there is a lot of issues.

First of all - I don't want to be offensive / rude, however wouldn't it be better to make multiple smaller commits and discuss before rewriting the "core" part of the project? It would save both of us time and wouldn't be >3000 lines for me to review. I hope you understand this.

1. One of the biggest change is the `Track` struct. 
 - I don't see the reason why should it be merged with `AudioFileInfo`. They both serve a different purpose - `AudioFileInfo` has info about the file itself - loaded metadata, it is an "input" to all the taggers. While `Track` is an "output" (in my implementation) - it has info already from the platforms itself. The issue with merging are:
 - Less readable code, a lot more `Option`s
 - Pretty much the struct is always going to have some fields empty (what I mean by this - if it's a result track from platform - it wouldn't have the format, path etc, and on the opposite - the loaded track from file would only have those few fields). This is more like code readability, maintainability, and also memory usage problem (since the `None` fields DO take up memory). 
 - The `fill_tags` function - yes I understand that it tries to parse the extra artists, remixers, info from the track title. However I do believe this is not mandatory and useless - each platform usually has own format for those fields (so it should be handled in individual platforms and then final `Track` object with clean data should be passed, like I did with Discogs), and also usually the platforms themselves already have fields in their APIs with those tags in clean format.
 - Bunch of code related issues - bunch of the `write_tags` tags are commented out for some reason, `platform` field in `Track` is an option, but there is `MusicPlatform::None` which would mean the same thing and would be easier to use, same goes for the `Vec<_>` fields - empty `Vec` is easier to deal with than `Option<Vec>` etc etc. But those could be fixed.
 - I do like the PlatformID structs (`BeatportID` and such), however they should be in their own platform implementation, or even better an `enum`. And also there are some mistakes such as Spotify IDs are Strings (or binary data if you will) rather than i64.
- The URLs, IDs tags - I guess some less repeated solution and no random letter prefixes would be better.

2. Another thing are `Helper` and `Matcher`.
- First of all those Regexes are kinda long and repeated, are you sure quantifiers couldn't be used there? Also you can probably match with ignored case (or just put the string to lowercase like I did), which would save the `[Ff]` and such parts in the regex. They are kinda unmaintainable, unreadable and janky.
- What was your reason for implementing own solution, rather than extending the existing ones? 
- From previous experience I do know that the `fuzzywuzzy` library is kinda unstable, doesn't fully support UTF-8 (so it can cause panics). In the original  `MatchingUtils` I used `strsim`, which does provide the exact same functionality (+ some extra), you could've just used that instead.
- Making `MATCH_ARTISTS` a static - with DiscogsTagger bascurtiz found out that not matching artists for Discogs is actually beneficial, compared to other platforms. So I'd keep it as normal parameter, so it can be disabled for individual platforms.

3. Bunch of minor unnecessary changes that don't make sense in some cases:
- `traxsource.rs` - renaming `title_elem` to `title_selector`, while the variable IS an element, not a selector. I don't get what you're trying to achieve doing that? Also adding comments such as `// Div selector (mutable to select the different divs in the HTML document)` - which is kinda unnecessary and not fully correct. 
- `junodownload.rs` - kinda the same issue as before - renaming variables for no reason / advantage, some even mislabeled.
- There are bunch more, but I am not sure which of them are intentional and which are artifact of running the formatter, and I don't wanna go over the small details right now.

4. `cargo fmt`
 - I see that you ran the `cargo fmt` or other code formatter. This is my bad for not running it myself, however I try to stick to the "official" code style as much as possible, only exceptions are long lines, where it makes sense. Running `cargo fmt` not only made the commit way bigger and harder to go thru, but also broke formatting of stuff like `audiofeatures.rs` (look at the commit diff to see what I am talking about), camelot key table unordered and such. So I wouldn't really want to run `cargo fmt` until this project is "ready" for it. I don't blame you for doing this, but yea I'd rather not do it now.


I hope what I said makes sense to you and also helps you understand, why I don't really want to merge this solution. I would really like to collaborate, but rather in smaller and agreeable chunks of code. Thank you.

aleixjf commented:
Hello Marekkon,

I wasn’t expecting you to merge this, but rather I wanted to “suggest” this approach. It makes full sense to do different branches/suggestions for each “new feature”, and it’s how I would usually work, but this was only for testing purposes, rather than implementing it and merging it to your build 😇

1. Indeed the major change is the Track structure!
The idea was to create a “common” structure for any object which contains metadata, no matter if the source was a local file or from an API/internet.
This should allow:
- Easier comparison of tracks in the Matcher
- “Shared” fill_tags function, implemented for the Track structure. (Despite maybe this could be approached/implemented as a Trait?)
Ideally, (on other languages) there would be a base class which could be later extended, but didn’t seem possible in Rust. (Though I’m having some ideas right now for a better approach)

However, I’ve got to be honest and confess that while I was rewriting some of your code I didn’t think about memory, and all this options really were a headache for rewriting ting the code + readability!

Indeed, the fill_tags function was written to get “extra” tags from the “available” ones! The idea was that, for example, if you got an artist string from the API (iTunes API), parse it and create the artists vector, so that the matcher could later compare the artists vector directly, which would be provided to the Track object either from the platform directly, or from this parse function after its creation.
At the end, the purpose of this is, once more, to help the matching system with this “extra” information!

Happy to read that you like the PlatformID structure, but yes, some of them may be wrong or incomplete (like I said, it’s not working and this was for testing purposes only!). The only one it’s properly written is the ITunesID one! ;)

About the URLs and the ID tags, the letters aren’t random ;)
In the ID3 tags, it says that user defined text fields should be/start TXXX, and user URLS should start with WXXX, or at least that’s how I understood it.

https://en.m.wikipedia.org › wiki › I...ID3 - Wikipedia

2. You’re right! I could match patterns with insensitive e case option, didn’t even think about it while writing the regexs! 
I didn’t know about strsim, so good to know! 👍🏻 
And nevermind the MATCH_ARTISTS static, it was a “temporal” solution for not messing with other things of the code!

3. Don’t know either why I renamed those variables, maybe I didn’t like the names? 👀
4. About cargo fmt, it was enabled by default when I installed the rust extensions on Visual Studio Code! If I do future suggestions/merge requests, I’ll be sure to have it disabled 👌🏻

Everything you said makes full sense to me, I hope that some of the clarifications above help you understand why I did some of the modifications I did!
And sure, I’d also love to collaborate, this project is very promising and I’ll try to help where possible! 🤓

Marekkon5 commented:
Hi, thank you for the quick reply. I appreciate your feedback and ideas, but I still stand in having separate "input" and "output" structs for the metadata. And yes trait would make it so both of them can have `fill_meta`. However I do believe `fill_meta` should be used on the input only - as I said the platforms themselves already provide clean data (in form of another field in the JSON or simillar), so they should have own implementations for this. Also matching against too much data will worsen the match rate. The goal of this project was to combine Beatport Tagger, Discogs Tagger and Traxsource Tagger into one, better solution. So the highest match rate is still priority. Hopefully we can find some "middle ground".

As for the URL tag - thanks for the info, didn't know that, however I am pretty sure the WXXX should be own tag and not under TXXX (like the current solution does), and I need to imement into the id3 API itself. However my point was more using something like single field for url and then just convert platform to string (instead of many ifs). 


I'll be away for the weekend, so I won't be able to collaborate / implement any of the items. However would be great if you could make some smaller PR with the iTunes improvements or simillar that I could actually merge. Thanks, I appreciate it.

BTW: Use rust analyzer extension for VS Code, it is faster and more updated.

Marekkon5 commented:
Hello, sorry for kinda ghosting this, but so much changed in 1T I will close it now because of inactivity and it's outdated now. Feel free to reopen/create/comment new issue/PR if you have any questions.

However if you want to continue - in the latest commits I've added a way to easily create custom platforms without needing to recompile OneTagger, and (with some pain it can in theory be used with any language). You can check it out at: https://github.com/Marekkon5/onetagger-platform-template

Thanks.
